crates/red_knot_python_semantic/src/suppression.rs

```diff
@@ -177,6 +140,116 @@ enum SuppressionTarget {
     Lint(LintId),
 }
 
+struct SuppressionsBuilder<'a> {
+    lint_registry: &'a LintRegistry,
+    source: &'a str,
+
+    /// `type: ignore` comments at the top of the file before any non-trivia code apply to the entire file.
+    /// This boolean tracks if there has been any non trivia token.
+    seen_non_trivia_token: bool,
+
+    line: Vec<Suppression>,
+    file: Vec<Suppression>,
+}
+
+impl<'a> SuppressionsBuilder<'a> {
+    fn new(source: &'a str, lint_registry: &'a LintRegistry) -> Self {
+        Self {
+            source,
+            lint_registry,
+            seen_non_trivia_token: false,
+            line: Vec::new(),
+            file: Vec::new(),
+        }
+    }
+
+    fn set_seen_non_trivia_token(&mut self) {
+        self.seen_non_trivia_token = true;
+    }
+
+    fn finish(mut self) -> Suppressions {
+        self.line.shrink_to_fit();
+        self.file.shrink_to_fit();
+
+        Suppressions {
+            file: self.file,
+            line: self.line,
+        }
+    }
+
+    fn add_comment(&mut self, comment: SuppressionComment, line_start: TextSize) {
+        let (suppressions, suppressed_range) =
+            // `type: ignore` comments at the start of the file apply to the entire range.
+            // > A # type: ignore comment on a line by itself at the top of a file, before any docstrings,
+            // > imports, or other executable code, silences all errors in the file.
+            // > Blank lines and other comments, such as shebang lines and coding cookies,
+            // > may precede the # type: ignore comment.
+            // > https://typing.readthedocs.io/en/latest/spec/directives.html#type-ignore-comments
+            if comment.kind.is_type_ignore() && !self.seen_non_trivia_token {
+                (
+                    &mut self.file,
+                    TextRange::new(0.into(), self.source.text_len()),
+                )
+            } else {
+                (
+                    &mut self.line,
+                    TextRange::new(line_start, comment.range.end()),
+                )
+            };
```

Comment:
This is the only new code in this block. Everything else was extracted from the `suppressions` functions